packages/resolve-url-loader/docs/advanced-features.md

```diff
@@ -1 +1,219 @@
+const myGenerator = (filename, uri, bases, isAbsolute, options) =>
+  isAbsolute ? [options.root] : [bases.subString, bases.value, bases.property, bases.selector];
+  
+const myJoinFn = createJoinFunction({
+  name     : 'myJoinFn',
+  scheme   : 'alstroemeria',
+  generator: myGenerator,
+  operation: defaultJoinOperation
+});
+```
+
+Notes
+
+* It's usually clearer for a simple `generator` to return `Array<string>`. Reserve using `function*` for where lazy evaluation is important.
+
+### How to: resolve to an asset that may not exist
+
+The default `operation` determines success based on whether a file exists in the webpack file system. If you wish to resolve a different file at the same location you will need to customise the `operation`.
+
+**example:** resolve a `.zip` file if the `.png` doesn't exist.
+
+```javascript
+const {
+  createJoinFunction,
+  defaultJoinGenerator,
+  webpackExistsSync
+} = require('resolve-url-loader');
+
+// accept a ".zip" file if the ".png" doesn't exist.
+const myOperation = (filename, uri, base, next, options) => {
+  const absolute    = path.normalize(path.join(base, uri));
+  const absoluteZip = absolute + '.zip';
+  return (
+    webpackExistsSync(options.fs, absolute) && options.fs.statSync(absolute).isFile() && absolute ||
+    webpackExistsSync(options.fs, absoluteZip) && options.fs.statSync(absoluteZip).isFile() && absoluteZip ||
+    next(absolute)
+  );
+};
+
+const myJoinFn = createJoinFunction({
+  name     : 'myJoinFn',
+  scheme   : 'alstroemeria',
+  generator: defaultJoinGenerator,
+  operation: myOperation
+});
+```
+
+Notes
+
+* You may combine `base` and `uri` arbitrarily since you control the "join".
+
+* You determine the predicate for success, but if the returned file does not exist then webpack will fail.
+
+* The webpack file-system is provided by the `enhanced-resolver-plugin` and does not contain `fs.existsSync`. Instead, use the `webpackExistsSync` utility function as shown.
+
+* Always `return` the result of calling `next()`. 
+
+### How to: perform a file-system search for an asset
+
+⚠️ **IMPORTANT** - This example is indicative only and is **not** advised.
+
+When this loader was originally relaesed it was very common for packages be broken to the point that a full file search was needed to locate assets referred to in CSS. While this was not performant some users really liked it.
+
+By customising the `generator` it is possibly to lazily provide a file search. So long as your criteria for success is that the file exists then the default `operation` can by used.
+
+**example:** search parent directories of the initial base path until you hit a package boundary
+
+```javascript
+const {
+  createJoinFunction,
+  defaultJoinOperation,
+  webpackExistsSync
+} = require('resolve-url-loader');
+
+// search up from the initial base path until you hit a package boundary
+const myGenerator = function* (filename, uri, bases, isAbsolute, options) {
```

Comment:
Not sure what your intended audience is, but personally I find the idea of writing a generator intimidating (almost as intimidating as configuring webpack 😅 ) - but it got me wondering whether it's possible to expose the api through functions that might get passed internally to a generator?

Reply:
In 99.5% of cases the function can return an `Array<string>` and need not be a generator. I think this is the most complicated example because it is lazy. If you want lazy then I think you need to work with the generator semantics.

As for whether we can do another layer of utility fns, I'm open to it with enough upvotes. 🤔